src/game/behaviors/door.inc.c

```diff
@@ -83,49 +84,56 @@ void bhv_door_loop(void) {
             play_warp_door_open_noise();
             break;
     }
-    if (o->oAction == DOOR_ACT_CLOSED) {
-        load_object_collision_model();
-    }
+
     bhv_door_rendering_loop();
 }
 
 void bhv_door_init(void) {
+    const f32 checkDist = 200.0f;
+
     f32 x = o->oPosX;
     f32 y = o->oPosY;
     f32 z = o->oPosZ;
-    struct Surface *floor;
 
-    find_room_floor(x, y, z, &floor);
-    if (floor != NULL) o->oDoorSelfRoom = floor->room;
+    o->oDoorSelfRoom = get_room_at_pos(x, y, z);
 
-    x = o->oPosX + sins(o->oMoveAngleYaw) *  200.0f;
-    z = o->oPosZ + coss(o->oMoveAngleYaw) *  200.0f;
+    x = o->oPosX + (sins(o->oMoveAngleYaw) *  checkDist);
+    z = o->oPosZ + (coss(o->oMoveAngleYaw) *  checkDist);
 
-    find_room_floor(x, y, z, &floor);
-    if (floor != NULL) o->oDoorForwardRoom = floor->room;
+    o->oDoorForwardRoom = get_room_at_pos(x, y, z);
 
-    x = o->oPosX + sins(o->oMoveAngleYaw) * -200.0f;
-    z = o->oPosZ + coss(o->oMoveAngleYaw) * -200.0f;
+    x = o->oPosX + (sins(o->oMoveAngleYaw) * -checkDist);
+    z = o->oPosZ + (coss(o->oMoveAngleYaw) * -checkDist);
 
-    find_room_floor(x, y, z, &floor);
-    if (floor != NULL) o->oDoorBackwardRoom = floor->room;
+    o->oDoorBackwardRoom = get_room_at_pos(x, y, z);
 
-    if (o->oDoorSelfRoom > 0 && o->oDoorSelfRoom < 60) {
-        gDoorAdjacentRooms[o->oDoorSelfRoom][0] = o->oDoorForwardRoom;
-        gDoorAdjacentRooms[o->oDoorSelfRoom][1] = o->oDoorBackwardRoom;
+    if (
+        // Ensure the room number is in bounds.
+        o->oDoorSelfRoom > 0 && o->oDoorSelfRoom < ARRAY_COUNT(gDoorAdjacentRooms)
+        // Only set gDoorAdjacentRooms for transition rooms.
+        && o->oDoorSelfRoom    != o->oDoorForwardRoom
+        && o->oDoorSelfRoom    != o->oDoorBackwardRoom
+        && o->oDoorForwardRoom != o->oDoorBackwardRoom
```

Comment:
Doors between rooms are intended to have small transition rooms under them so that objects (and level geometry) in both of the non-transition rooms on each side of the door are loaded when Mario is near (in the transition room). gDoorAdjacentRooms is used to store the adjacent rooms for each of those transition rooms.

In the past, if a doors that was between two rooms but didn't have a transition room loaded the objects in both rooms at once, that would have been a side effect of the bug that this fixes (doors setting non-transition rooms and overwriting transition room data), and objects outside the room would have been always loaded/visible regardless of Mario's proximity to the door (and would have been a little buggy in rooms with multiple doors).

A possible solution would be to get rid of transition rooms from the level collision and geolayouts entirely and link the adjacent rooms data with the door itself, and only load both sides whenever the door is open, but that would likely mean every object every frame would have to loop through wherever data that is stored to check if its room is loaded (cur_obj_is_mario_in_room), so loaded rooms would have to be changed from a single "current loaded room" to an array of booleans for all loaded rooms, and that loading the level geometry of each room would also need to use that system, since currently transition rooms just have a copy of both rooms' geolayouts, and geolayout switches can't load multiple rooms at once. Definitely out of 2.1 scope though.

Reply:
what you're saying does make sense but at the same time you could still have situations where the creator didn't have the middle room and doesn't want the transition to be buggy OR always wants that room to be loaded from the other room and intended it to be that way, think like a basic hallway with doors between two rooms. that hallway could be always rendered in either room for an easy seamless transition without much fill (especially without camera collision), but you wouldn't want one room rendering the other room